src/en/news/blog/2023/reef-osds-per-nvme/index.md

```diff
@@ -0,0 +1,137 @@
+
+```paste <(cat /proc/cpuinfo | grep "core id") <(cat /proc/cpuinfo | grep "processor") | sed 's/[[:blank:]]/ /g'```
+
+This showed that processors 0-63 corresponded to physical cores and processors 64-127 corresponded to HT cores.  Next, numactl was invoked to utilize an equal number of physical and HT processors (which for this article we'll refer to generally as CPU threads).  For instance, to assign OSDs to a pool of 12 physical and 12 HT cores (for a total of 24 cpu threads, or 4 threads per NVMe), OSDs were invoked thusly:
+
+```...numactl --physcpubind=0-12,64-76 /usr/local/bin/ceph-osd```
+
+Tests were conducted with between 2 and 20 CPU Threads per NVMe drive.  FIO was configured to first pre-fill RBD volume(s) with large writes, followed by 4MB and 4KB IO tests for 60 seconds each.  Certain background processes, such as scrub, deep scrub, pg autoscaling, and pg balancing were disabled.  An RBD pool with a static 16384 PGs (higher than typically recommended) and 3x replication was used.
+
+# 4MB RBD Read and Write Throughput 
+
+![](images/Ceph_Reef_-_FIO_4MB_Read_Throughput.svg) 
+![](images/Ceph_Reef_-_FIO_4MB_Write_Throughput.svg)    
+
+When I started this testing I wasn't expecting to see a significant performance difference with multiple OSDs per NVMe device for large IO workloads.  For reads, the difference is fairly small, though the 2 OSD configuration showed a fairly consistent small advantage.  For writes however, I was surprised to observe that there was a moderate throughput slump that peaked at 12 CPU threads per NVMe versus the single OSD configuration.
+
+![](images/Ceph_Reef_-_FIO_4MB_2_OSD_NVMe_Throughput_Advantage.svg)
+
+These were short running tests, but the cluster was recreated at every CPU count sample point and we see a clear trend over multiple samples.  The effect is likely to be real and appears to match the behavior we saw in Octopus and Pacific where 4MB Write throughput was actually lower in the 2 OSD/NVMe configuration.  Unfortunately I do not yet have an explanation for this behavior.
+
+# 4KB RBD Random Read and Write IOPS
+
+![](images/Ceph_Reef_-_FIO_4KB_Random_Read_IOPS.svg)
+![](images/Ceph_Reef_-_FIO_4KB_Random_Write_IOPS.svg)
+![](Ceph_Reef_-_FIO_4KB_2_OSD_NVMe_IOPS_Advantage.svg)
+
+Back during Nautilus we saw a significant advantage with 2 OSDs/NVMe in both 4KB random read and random write tests.  What we are seeing now in Reef is that there is no particular IOPS advantage to deploying 2 OSDs on each NVMe drive unless you have 16 CPU threads per NVMe drive or more.  In fact for deployments that have 2-4 CPU threads per NVMe, there's a slight IOPS hit.  There's a couple of immediate take aways from this testing:
```

Comment:
With Nautilus we saw
takeaways

Also, might the kernel / OS version between Stream 8 and whatever you used in the past factor in?

Reply:
The tests that were run for Nautilus/Octopus/Pacific were all done at the same time on the same HW/OS.  They are absolutely on a different platform vs the Reef tests we are performing now, but I believe the effect that we saw in the previous tests is real.  I can try to clarify that if you think it's valuable?

Reply:
No, probably not worth the trouble, but I was wondering for my own purposes as well.  I'm in the middle of trying to figure out why updating the kernel on Aerospike systems with NVMe drives increased CPU utilization and resulted in more errors, so it's top of mind.